File: packages/api/src/messaging/channel.ts
/**
 * Channel management utilities for RabbitMQ
 */
import { Channel, Options } from 'amqplib';
import { ConnectionPool } from './connection';

export class ChannelManager {
  private connectionPool: ConnectionPool;
  private channels: Map<string, Channel> = new Map();

  constructor(connectionPool: ConnectionPool) {
    this.connectionPool = connectionPool;
  }

  /**
   * Get or create a channel
   */
  async getChannel(channelId: string = 'default'): Promise<Channel> {
    // Check if we have an existing channel
    const existingChannel = this.channels.get(channelId);
    if (existingChannel) {
      return existingChannel;
    }

    // Create new channel
    try {
      const connection = await this.connectionPool.getConnection();
      const channel = await connection.createChannel();

      // Handle channel errors
      channel.on('error', (err) => {
        console.error(`Channel ${channelId} error:`, err);
        this.channels.delete(channelId);
      });

      channel.on('close', () => {
        console.warn(`Channel ${channelId} closed`);
        this.channels.delete(channelId);
      });

      this.channels.set(channelId, channel);
      console.log(`Created new channel: ${channelId}`);
      return channel;
    } catch (error) {
      console.error(`Failed to create channel ${channelId}:`, error);
      throw new Error(`Could not create channel: ${error}`);
    }
  }

  /**
   * Close a specific channel
   */
  async closeChannel(channelId: string = 'default'): Promise<void> {
    const channel = this.channels.get(channelId);
    if (channel) {
      try {
        console.log(`Closing channel: ${channelId}`);
        await channel.close();
      } catch (error) {
        console.error(`Error closing channel ${channelId}:`, error);
      } finally {
        this.channels.delete(channelId);
      }
    }
  }

  /**
   * Close all managed channels
   */
  async closeAllChannels(): Promise<void> {
    const channelIds = Array.from(this.channels.keys());
    await Promise.all(channelIds.map((id) => this.closeChannel(id)));
  }

  /**
   * Declare an exchange
   */
  async declareExchange(
    channel: Channel,
    exchangeName: string,
    exchangeType: 'direct' | 'topic' | 'fanout' | 'headers' = 'topic',
    options: Options.AssertExchange = {}
  ): Promise<void> {
    try {
      await channel.assertExchange(exchangeName, exchangeType, {
        durable: true,
        ...options,
      });
      console.log(`Declared exchange: ${exchangeName} (type: ${exchangeType})`);
    } catch (error) {
      console.error(`Failed to declare exchange ${exchangeName}:`, error);
      throw error;
    }
  }

  /**
   * Declare a queue
   */
  async declareQueue(
    channel: Channel,
    queueName: string,
    options: Options.AssertQueue = {}
  ): Promise<{ queue: string; messageCount: number; consumerCount: number }> {
    try {
      const result = await channel.assertQueue(queueName, {
        durable: true,
        ...options,
      });
      console.log(`Declared queue: ${queueName}`);
      return result;
    } catch (error) {
      console.error(`Failed to declare queue ${queueName}:`, error);
      throw error;
    }
  }

  /**
   * Bind a queue to an exchange with a routing key
   */
  async bindQueue(
    channel: Channel,
    queueName: string,
    exchangeName: string,
    routingKey: string
  ): Promise<void> {
    try {
      await channel.bindQueue(queueName, exchangeName, routingKey);
      console.log(
        `Bound queue ${queueName} to exchange ${exchangeName} with routing key ${routingKey}`
      );
    } catch (error) {
      console.error(
        `Failed to bind queue ${queueName} to exchange ${exchangeName}:`,
        error
      );
      throw error;
    }
  }

  /**
   * Set prefetch count for a channel (QoS)
   */
  async setPrefetch(channel: Channel, count: number): Promise<void> {
    try {
      await channel.prefetch(count);
      console.log(`Set prefetch count to ${count}`);
    } catch (error) {
      console.error(`Failed to set prefetch count:`, error);
      throw error;
    }
  }
}

File: packages/agents/src/messaging/channel.py
"""Channel management utilities for RabbitMQ"""
import logging
from typing import Optional, Dict
from contextlib import contextmanager
import pika
from pika.channel import Channel
from pika.exceptions import AMQPChannelError

from .connection import ConnectionPool

logger = logging.getLogger(__name__)


class ChannelManager:
    """Manages RabbitMQ channels with automatic recovery"""

    def __init__(self, connection_pool: ConnectionPool):
        """
        Initialize channel manager

        Args:
            connection_pool: Connection pool to use for creating channels
        """
        self.connection_pool = connection_pool
        self._channels: Dict[str, Channel] = {}

    def get_channel(self, channel_id: str = "default") -> Channel:
        """
        Get or create a channel

        Args:
            channel_id: Identifier for the channel

        Returns:
            Active RabbitMQ channel

        Raises:
            AMQPChannelError: If channel creation fails
        """
        # Check if we have an existing open channel
        if channel_id in self._channels:
            channel = self._channels[channel_id]
            if channel.is_open:
                return channel
            else:
                logger.warning(f"Channel {channel_id} is closed, creating new one")
                del self._channels[channel_id]

        # Create new channel
        try:
            connection = self.connection_pool.get_connection()
            channel = connection.channel()
            self._channels[channel_id] = channel
            logger.info(f"Created new channel: {channel_id}")
            return channel

        except Exception as e:
            logger.error(f"Failed to create channel {channel_id}: {e}")
            raise AMQPChannelError(f"Could not create channel: {e}")

    def close_channel(self, channel_id: str = "default") -> None:
        """
        Close a specific channel

        Args:
            channel_id: Identifier for the channel to close
        """
        if channel_id in self._channels:
            channel = self._channels[channel_id]
            try:
                if channel.is_open:
                    logger.info(f"Closing channel: {channel_id}")
                    channel.close()
            except Exception as e:
                logger.error(f"Error closing channel {channel_id}: {e}")
            finally:
                del self._channels[channel_id]

    def close_all_channels(self) -> None:
        """Close all managed channels"""
        channel_ids = list(self._channels.keys())
        for channel_id in channel_ids:
            self.close_channel(channel_id)

    @contextmanager
    def channel_context(self, channel_id: str = "default"):
        """
        Context manager for channel usage

        Args:
            channel_id: Identifier for the channel

        Yields:
            Active RabbitMQ channel

        Example:
            with channel_manager.channel_context("my_channel") as channel:
                channel.basic_publish(...)
        """
        channel = self.get_channel(channel_id)
        try:
            yield channel
        except Exception as e:
            logger.error(f"Error in channel context {channel_id}: {e}")
            # Close and remove the channel on error
            self.close_channel(channel_id)
            raise

    def declare_exchange(
        self,
        channel: Channel,
        exchange_name: str,
        exchange_type: str = "topic",
        durable: bool = True,
    ) -> None:
        """
        Declare an exchange

        Args:
            channel: Channel to use for declaration
            exchange_name: Name of the exchange
            exchange_type: Type of exchange (direct, topic, fanout, headers)
            durable: Whether the exchange survives broker restart
        """
        try:
            channel.exchange_declare(
                exchange=exchange_name,
                exchange_type=exchange_type,
                durable=durable,
            )
            logger.info(f"Declared exchange: {exchange_name} (type: {exchange_type})")
        except Exception as e:
            logger.error(f"Failed to declare exchange {exchange_name}: {e}")
            raise

    def declare_queue(
        self,
        channel: Channel,
        queue_name: str,
        durable: bool = True,
        exclusive: bool = False,
        auto_delete: bool = False,
        arguments: Optional[Dict] = None,
    ) -> pika.frame.Method:
        """
        Declare a queue

        Args:
            channel: Channel to use for declaration
            queue_name: Name of the queue
            durable: Whether the queue survives broker restart
            exclusive: Whether the queue is exclusive to this connection
            auto_delete: Whether the queue is deleted when last consumer unsubscribes
            arguments: Additional queue arguments (e.g., TTL, dead letter exchange)

        Returns:
            Queue declaration result
        """
        try:
            result = channel.queue_declare(
                queue=queue_name,
                durable=durable,
                exclusive=exclusive,
                auto_delete=auto_delete,
                arguments=arguments or {},
            )
            logger.info(f"Declared queue: {queue_name}")
            return result
        except Exception as e:
            logger.error(f"Failed to declare queue {queue_name}: {e}")
            raise

    def bind_queue(
        self,
        channel: Channel,
        queue_name: str,
        exchange_name: str,
        routing_key: str,
    ) -> None:
        """
        Bind a queue to an exchange with a routing key

        Args:
            channel: Channel to use for binding
            queue_name: Name of the queue
            exchange_name: Name of the exchange
            routing_key: Routing key pattern
        """
        try:
            channel.queue_bind(
                queue=queue_name,
                exchange=exchange_name,
                routing_key=routing_key,
            )
            logger.info(
                f"Bound queue {queue_name} to exchange {exchange_name} "
                f"with routing key {routing_key}"
            )
        except Exception as e:
            logger.error(
                f"Failed to bind queue {queue_name} to exchange {exchange_name}: {e}"
            )
            raise

    def __enter__(self) -> "ChannelManager":
        """Context manager entry"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        """Context manager exit"""
        self.close_all_channels()
